### PR TITLE
Provide caret rect to text input plugin

### PR DESCRIPTION
**Pull request by knopp**
https://github.com/flutter/flutter/issues/77545

Adds required code to provide current caret rectangle to text input plugin. This rect, with addition to current editable transform can be used to position the accent selection menu on macOS similarly to how IME popup is positioned on iOS.

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt.
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

**Comment by knopp** [comment]
https://github.com/flutter/flutter/issues/77052

**Comment by knopp** [comment]
https://github.com/flutter/flutter/issues/30660

File: packages/flutter/lib/src/services/text_input.dart
Comment (knopp) [reply packages/flutter/lib/src/services/text_input.dart:1038]: I wasn't trying to match CGRectNull (i think that's [inf, inf, 0, 0]), just to be consistent with `setComposingRect`.

File: packages/flutter/lib/src/widgets/editable_text.dart
Comment (knopp) [reply packages/flutter/lib/src/widgets/editable_text.dart]: Addressed [here](https://github.com/flutter/flutter/pull/77608/commits/a652c0324adefdae2d03fb88dd2d87e0065f3a2c).

Comment (LongCatIsLooong) [reply packages/flutter/lib/src/widgets/editable_text.dart]: Looks like we are skipping reporting the caret rect to the engine when there's no caret. Does the engine need to know that the caret is gone (instead of using the out-of-date caret rect that was previously sent to it)?

Comment (knopp) [reply packages/flutter/lib/src/widgets/editable_text.dart]: The purpose of caret rect is to position the IME candidate popup. The popup is displayed after long pressing a key (or possibly while typing with some languages). So if there is any selection it will be replaced by a character before the popup is displayed. I'm not aware of a way to get the IME popup visible without caret present. And should that ever happen, last caret position seems to me like a good enough fallback.

Comment (LongCatIsLooong) [reply packages/flutter/lib/src/widgets/editable_text.dart]: Wouldn't the accent popup show up at the wrong location if you:
1. change the selection to somewhere in the document that's far away from the current caret location (the framework won't send anything in this case because it's a non-collapsed selection).
2. long press a key to trigger the popup

**Comment by LongCatIsLooong** [comment]
Looks like @cbracken is about to land a patch that addresses something similar. Maybe we should wait for that patch to land?

I'm not familiar with how accents work on macos.  Does long-pressing a key initiate a composing region (that highlights the base character) on macos?

File: packages/flutter/lib/src/widgets/editable_text.dart
Comment (knopp) [reply packages/flutter/lib/src/widgets/editable_text.dart]: Long pressing key first clears the selection replacing it with pressed character, moving caret to new position. I don't think this would cause any issues. The way this works is that the popup shows the moment character would normally start repeating (after a long key press).

**Comment by knopp** [comment]
> Looks like @cbracken is about to land a patch that addresses something similar. Maybe we should wait for that patch to land?

Sure. I wasn't aware @cbracken was working on this when I opened the PR. Don't mean to step on anyone's toes :)

> I'm not familiar with how accents work on macos. Does long-pressing a key initiate a composing region (that highlights the base character) on macos?

Not necessarily. The text input controller suppresses normal  key-repeat behavior, showing selection popup on first key repeat instead. However it only enters compositing mode after you change the character selection within the popup. Which currently doesn't work in flutter due to way caret movement and arrow events are handled (but you can still pick up option directly by pressing a number, which works). In which case it generates text insertion event, skipping the composition step altogether.

image

**Review by cbracken — COMMENTED** [review]
Can you provide a bit more context on why this is required? `TextInput.setMarkedTextRect` should provide the caret position when not composing. https://github.com/flutter/flutter/blob/48f5674085b72bb3d139a056394642d2d6670f3f/packages/flutter/lib/src/widgets/editable_text.dart#L2442-L2463

I'm not aware of any cases where both an IME candidate menu and the accent pop-up are in use at the same time -- in local testing with Japanese and Korean text, I get direct input of the vowel in question (e.g. あ) and no accent menu.

**Comment by knopp** [comment]
@cbracken, I did try `setMarkedTextRect` first, but without compositing it keeps sending `Rect.fromLTRB(-1.0, 0.0, 1.0, 19.0)` regardless of where the caret is. If that's not correct behavior perhaps `setMarkedTextRect` should be fixed?

**Comment by cbracken** [comment]
> @cbracken, I did try `setMarkedTextRect` first, but without compositing it keeps sending `Rect.fromLTRB(-1.0, 0.0, 1.0, 19.0)` regardless of where the caret is.

Ah interesting; yep looking at [this line](https://github.com/flutter/flutter/blob/48f5674085b72bb3d139a056394642d2d6670f3f/packages/flutter/lib/src/widgets/editable_text.dart#L2455), `TextRange.isValid` returns false when start/end are -1. 

@LongCatIsLooong any concerns with the general approach? From my point of view this seems like a reasonable message to add, and I think this actually solves an additional issue with CJK candidate menu positioning. On some platforms, the candidate dropdown movies with the caret, on others it remains positioned at the bottom left corner of the composing rect. Currently on platforms that track the caret, we track the bottom-right corner of the composing rect as a proxy for that.

An alternative would be to have the `setMarkedTextRect` message pass the caret even when not composing, but given the message name and likely resulting confusion, I'd really prefer the current approach.

File: packages/flutter/lib/src/widgets/editable_text.dart
Comment (LongCatIsLooong) [reply packages/flutter/lib/src/widgets/editable_text.dart]: The communication between the text input plugin and the framework is asynchronous. When the software keyboard clears the selection and changes the caret's position, the text input plugin needs to query the new caret coordinates from the framework. There seems to be a chance that the framework can't respond in time, resulting in the outdated caret coordinates being reported in the `firstRect` method?

Comment (knopp) [reply packages/flutter/lib/src/widgets/editable_text.dart]: I'm not entirely sure what do you mean by software keyboard? if you mean mean the selection popup, that is shown as response to previous keypress (that would clear the selection). The time between the initial key press and showing popup (when long pressing the key) is about a second, plenty of time for framework to notify plugin of cursor location.

**Review by justinmc — COMMENTED** [review]
I'll defer to @LongCatIsLooong on this, but it seems to make sense to me.

File: packages/flutter/lib/src/widgets/editable_text.dart
Comment (knopp) [reply packages/flutter/lib/src/widgets/editable_text.dart:2475]: TextInputConnection checks if the rect is same, and so skips the message. 

I'm not quite sure how infinite loop could happen, it is started in `_openInputConnection` and repeated every frame for as long as there is input connection. It is exactly the same behavior as `_updateComposingRectIfNeeded`, so if that doesn't cause infinite loops neither should this.

File: packages/flutter/lib/src/widgets/editable_text.dart
Comment (LongCatIsLooong) [reply packages/flutter/lib/src/widgets/editable_text.dart]: ah sorry forgot this is macOS & hardware keyboard, I meant the key event. Sounds good.

File: packages/flutter/lib/src/widgets/editable_text.dart
Comment (justinmc) [reply packages/flutter/lib/src/widgets/editable_text.dart:2475]: You're right, _updateComposingRectIfNeeded does seem to do the same thing so I think it's ok 👍

**Review by LongCatIsLooong — APPROVED** [review]
LGTM modulo comments from other reviewers

File: packages/flutter/lib/src/services/text_input.dart
Comment (cbracken) [inline packages/flutter/lib/src/services/text_input.dart:1032]: nit: `s/only used/used/` just in case we do start using it in another embedder at some point.

Comment (cbracken) [reply packages/flutter/lib/src/services/text_input.dart:1032]: Sent a fix for the othe case in https://github.com/flutter/flutter/pull/77951

**Comment by goderbauer** [comment]
(PR triage) @LongCatIsLooong @justinmc @knopp What's the status of this PR?

**Comment by knopp** [comment]
I rebased the PR and added unit tests (as requested by @justinmc). Anything else needed here?

**Review by justinmc — APPROVED** [review]
LGTM 👍 

The build is currently broken, but this can be merged when it's green.

**Pull request merged**
